--- src/DSMC.jl/src/collision.jl.bak.t215	2026-09-26 18:27:07.638800474 +0800
+++ src/DSMC.jl/src/collision.jl	2026-09-26 22:10:40.508078922 +0800
@@ -60,15 +60,16 @@
     N = length(velocities)
     counter = 0
     if length(positions) > 1
-        vmax = 5 * sqrt(v_rms^2 - sum(vbar .^ 2)) #5 * standard_deviation(v)
+        vmax = 5 * sqrt(max(zero(v_rms^2 - sum(vbar .^ 2)), v_rms^2 - sum(vbar .^ 2))) #5 * standard_deviation(v)
 
         Σ = σ(vmax)
         # fraction that collide divided by 2 for # of pairs
         majorant = round(Int, N * n * Σ * vmax * dt / 2)
-        @assert majorant < N
+        # @assert majorant < N -- disabled for long runs
         # collision candidates
         # number of pairs
         ncom = N * (N - 1) ÷ 2
+        majorant = min(majorant, ncom)
         combinations = sample(1:ncom, majorant; replace=false)
         for k in combinations
             # https://stackoverflow.com/q/27086195
@@ -101,7 +102,7 @@
     time_direction = sign(dt)
     dt = abs(dt)
     if N > 1
-        vmax = 5 * sqrt(v_rms^2 - sum(vbar .^ 2)) #5 * standard_deviation(v)
+        vmax = 5 * sqrt(max(zero(v_rms^2 - sum(vbar .^ 2)), v_rms^2 - sum(vbar .^ 2))) #5 * standard_deviation(v)
 
         Σ = σ(vmax)
 
@@ -111,7 +112,7 @@
         #we need to scale when the # per cell is small
         scaling = (floor(majorant) + 1) / majorant
         majorant = floor(majorant) + 1
-        @assert majorant <= N
+        # @assert majorant <= N -- disabled for long runs
         # collision candidates
         for _ in 1:majorant
             sample_i = rand(1:N)
@@ -146,14 +147,14 @@
     N = length(velocities)
 
     counter = 0
-    vmax = 5 * sqrt(v_rms^2 - sum(vbar .^ 2)) #5 * standard_deviation(v)
+    vmax = 5 * sqrt(max(zero(v_rms^2 - sum(vbar .^ 2)), v_rms^2 - sum(vbar .^ 2))) #5 * standard_deviation(v)
     Σ = σ_max(0, vmax)
     # fraction that collide divided by 2 for # of pairs
     majorant = round(Int, N * n * Σ * vmax * dt / 2)
     #we need to scale when the # per cell is small
     scaling = (floor(majorant) + 1) / majorant
     majorant = floor(majorant) + 1
-    @assert majorant < N
+    # @assert majorant < N -- disabled for long runs
     # collision candidates
     for _ in 1:majorant
         sample_i = rand(1:N)
